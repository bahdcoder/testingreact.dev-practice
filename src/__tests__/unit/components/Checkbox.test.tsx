import React from 'react'
import { axe } from 'jest-axe'
import { render, fireEvent } from '@testing-library/react'

import Checkbox from '../../../components/Checkbox'

describe('The <Checkbox /> component', () => {
  const defaultCheckboxProps = {
    label: 'TEST_CHECKBOX_LABEL',
    id: 'TEST_CHECKBOX_ID',
    background: '#000',
    checkMarkBackground: '#fff',
    onChange: jest.fn(),
    checked: false,
  }

  const setupCheckbox = (props = defaultCheckboxProps) =>
    render(<Checkbox {...props} />)

  it('should set the id of the checkbox correctly', () => {
    const { getByLabelText } = setupCheckbox()

    expect(getByLabelText(defaultCheckboxProps.label).id).toBe(
      defaultCheckboxProps.id,
    )
  })

  it('renders correctly', () => {
    const { asFragment } = setupCheckbox()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should set the checkbox to the correct checked state', () => {
    const { getByLabelText } = setupCheckbox({
      ...defaultCheckboxProps,
      checked: true,
    })

    expect(getByLabelText(defaultCheckboxProps.label)).toBeChecked()
  })

  it('should call the onChange handler when checkbox is clicked', () => {
    const { getByLabelText } = setupCheckbox({
      ...defaultCheckboxProps,
      checked: true,
    })

    const checkbox = getByLabelText(defaultCheckboxProps.label)

    fireEvent.click(checkbox)

    expect(defaultCheckboxProps.onChange).toHaveBeenCalled()
  })

  // The main test here is: Form elements should have labels.
  // Demonstrate by adding it.only to this test, and
  // removing the label from the checkbox component.
  it('should not fail any accessibility tests', async () => {
    const { container } = setupCheckbox()

    expect(await axe(container)).toHaveNoViolations()
  })
})
