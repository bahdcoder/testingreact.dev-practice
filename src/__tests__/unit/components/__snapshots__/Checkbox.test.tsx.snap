// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Checkbox /> component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm StDqN"
    data-testid="Checkbox"
  >
    <input
      class="sc-AxhCb iqvdaa"
      id="TEST_CHECKBOX_ID"
      type="checkbox"
    />
    <label
      class="sc-AxiKw cwRuvq"
      for="TEST_CHECKBOX_ID"
    >
      <span
        class="sc-AxirZ bZbnmP"
      >
        TEST_CHECKBOX_LABEL
      </span>
    </label>
  </div>
</DocumentFragment>
`;
